src/core/validation/types.ts

```diff
@@ -0,0 +1,62 @@
+import { Falsy } from "../../utils";
+import { FieldDescriptor } from "../types/field-descriptor";
+
+/**
+ * Function responsible for validating single field.
+ *
+ * @param value - value to be validated of type `T`
+ *
+ * @returns validation error of type `Err`, or `null` when field is valid
+ */
+export type Validator<T, Err> =
+  | Validator.Sync<T, Err>
+  | Validator.Async<T, Err>;
```

Comment:
async validation is planned for post-mvp, if it will be causing issues we can leave it out for now